File: Server/knight-online/shared/src/main/java/com/knightonline/shared/data/enums/NationEnum.java
package com.knightonline.shared.data.enums;

import java.io.Serializable;
import java.util.HashMap;
import java.util.Map;

/**
 * @author Mamaorha
 *
 */
public enum NationEnum implements Serializable
{
	NO_NATION(0), KARUS(1), ELMORAD(2);
	
	private Integer value;
	private static Map<Integer, NationEnum> namesMap = new HashMap<>();

	NationEnum(Integer value)
	{
		this.value = value;
	}

	public Integer getValue()
	{
		return this.value;
	}
	
	public static NationEnum forValue(Integer value)
	{
		return namesMap.get(value);
	}
	
	static
	{
		final NationEnum[] values = NationEnum.values();
		
		for (NationEnum temp : values)
		{
			namesMap.put(temp.getValue(), temp);
		}
	}
}

File: Client/WarFare/NationEnum.cpp
#include "NationEnum.h"

NationEnum & NationEnum::instance()
{
	static NationEnum INSTANCE;
	return INSTANCE;
}

NationEnum::NationEnum() {}

NationEnum::NationEnum(char * name, short value) : Enum(name, value)
{
	instance().values[value] = this;
}

NationEnum & NationEnum::forValue(short value)
{
	Enum * enumType = instance().forValueT(value);

	return *(NationEnum *)enumType;
}

const NationEnum & NationEnum::NO_NATION = NationEnum("NO_NATION", 0);
const NationEnum & NationEnum::KARUS = NationEnum("KARUS", 1);
const NationEnum & NationEnum::ELMORAD = NationEnum("ELMORAD", 2);